src/raw/mod.rs

```diff
@@ -0,0 +1,1849 @@
+This allows accessing nested documents without reallocation.  [Doc]
+must always be accessed via a pointer type, similarly to `[T]` and `str`.
+
+This type will coexist with the now deprecated [DocRef] type for at
+least one minor release.
+
+The below example constructs a bson document in a stack-based array,
+and extracts a &str from it, performing no heap allocation.
+
+```rust
+use bson::raw::Doc;
+
+let bytes = b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00";
+assert_eq!(Doc::new(bytes)?.get_str("hi")?, Some("y'all"));
+# Ok::<(), bson::raw::RawError>(())
+```
+
+### Iteration
+
+[`Doc`] implements [`IntoIterator`](std::iter::IntoIterator), which can also
+be accessed via [`DocBuf::iter`].
+
+```rust
+use bson::doc;
+use bson::raw::{DocBuf, elem::Element};
+
+let doc = DocBuf::from_document(&doc! {"crate": "rawbson", "license": "MIT"});
+let mut dociter = doc.iter();
+
+let (key, value): (&str, Element) = dociter.next().unwrap()?;
+assert_eq!(key, "crate");
+assert_eq!(value.as_str()?, "rawbson");
+
+let (key, value): (&str, Element) = dociter.next().unwrap()?;
+assert_eq!(key, "license");
+assert_eq!(value.as_str()?, "MIT");
+# Ok::<(), bson::raw::RawError>(())
+```
+
+### serde support
+
+There is also serde deserialization support.
+
+Serde serialization support is not yet provided.  For now, use
+[`bson::to_document`] instead, and then serialize it out using
+[`bson::Document::to_writer`] or [`DocBuf::from_document`].
+
+```rust
+use serde::Deserialize;
+use bson::{doc, Document, oid::ObjectId, DateTime};
+use bson::raw::{DocBuf, de::from_docbuf};
+
+#[derive(Deserialize)]
+#[serde(rename_all="camelCase")]
+struct User {
+    #[serde(rename = "_id")]
+    id: ObjectId,
+    first_name: String,
+    last_name: String,
+    birthdate: Option<chrono::DateTime<chrono::Utc>>,
+    #[serde(flatten)]
+    extra: Document,
+}
+
+let doc = DocBuf::from_document(&doc!{
+    "_id": ObjectId::with_string("543254325432543254325432")?,
+    "firstName": "John",
+    "lastName": "Doe",
+    "birthdate": null,
+    "luckyNumbers": [3, 60, 2147483647],
+    "nickname": "Red",
+});
+
+let user: User = from_docbuf(&doc)?;
+assert_eq!(user.id.to_hex(), "543254325432543254325432");
+assert_eq!(user.first_name, "John");
+assert_eq!(user.last_name, "Doe");
+assert_eq!(user.extra.get_str("nickname")?, "Red");
+assert!(user.birthdate.is_none());
+# Ok::<(), Box<dyn std::error::Error>>(())
+```
+*/
+
+use std::{
+    borrow::Borrow,
+    convert::{TryFrom, TryInto},
+    ops::Deref,
+};
+
+use chrono::{DateTime, Utc};
+
+#[cfg(feature = "decimal128")]
+use crate::decimal128::Decimal128;
+
+use crate::{document::ValueAccessError, oid, spec::ElementType, Bson};
+
+pub mod de;
+pub mod elem;
+
+#[cfg(test)]
+mod props;
+
+/// Error to indicate that either a value was empty or it contained an unexpected
+/// type, for use with the direct getters.
+#[derive(Debug, PartialEq)]
+pub enum RawError {
```

Comment:
It would be nice to merge this with the primary crate Error type. The main difference is needing to be able to signal malformed bsom when accessing elements.